eacl/record.go

```diff
@@ -1,325 +1,212 @@
-			filters[i] = *r.filters[i].ToV2()
+	filters := m.GetFilters()
+	if filters != nil {
+		r.filters = make([]Filter, len(filters))
+		for i := range filters {
+			err := r.filters[i].readFromV2(filters[i])
+			if err != nil {
+				return fmt.Errorf("invalid filter #%d: %w", i, err)
+			}
 		}
-
-		v2.SetFilters(filters)
+	} else {
+		r.filters = nil
 	}
 
-	v2.SetAction(r.action.ToV2())
-	v2.SetOperation(r.operation.ToV2())
+	r.action = Action(m.GetAction())
+	r.operation = acl.Op(m.GetOperation())
 
-	return v2
-}
-
-// NewRecord creates and returns blank Record instance.
-//
-// Defaults:
-//   - action: ActionUnknown;
-//   - operation: OperationUnknown;
-//   - targets: nil,
-//   - filters: nil.
-func NewRecord() *Record {
-	return new(Record)
-}
-
-// CreateRecord creates, initializes with parameters and returns Record instance.
-func CreateRecord(action Action, operation Operation) *Record {
-	r := NewRecord()
-	r.action = action
-	r.operation = operation
-	r.targets = []Target{}
-	r.filters = []Filter{}
-
-	return r
+	return nil
 }
 
-// NewRecordFromV2 converts v2 acl.EACLRecord message to Record.
-func NewRecordFromV2(record *v2acl.Record) *Record {
-	r := NewRecord()
+// writeToV2 writes Record to [v2acl.Target] messages of the NeoFS API protocol.
+func (r Record) writeToV2(m *v2acl.Record) {
+	if r.filters != nil {
+		filters := make([]v2acl.HeaderFilter, len(r.filters))
+		for i := range r.filters {
+			r.filters[i].writeToV2(&filters[i])
+		}
 
-	if record == nil {
-		return r
+		m.SetFilters(filters)
+	} else {
+		m.SetFilters(nil)
 	}
 
-	r.action = ActionFromV2(record.GetAction())
-	r.operation = OperationFromV2(record.GetOperation())
-
-	v2targets := record.GetTargets()
-	v2filters := record.GetFilters()
-
-	r.targets = make([]Target, len(v2targets))
-	for i := range v2targets {
-		r.targets[i] = *NewTargetFromV2(&v2targets[i])
+	m.SetTargets(r.target.toV2())
+	m.SetAction(v2acl.Action(r.action))
+	m.SetOperation(v2acl.Operation(r.operation))
+}
+
+// IsForRole checks whether the access rule applies to the given role or not.
+// Role MUST be one of the supported enum values.
+func (r Record) IsForRole(role Role) bool {
+	switch role {
+	default:
+		panic(fmt.Sprintf("unsupported value of enum %T: %v", role, role))
+	case
+		RoleContainerOwner,
+		RoleSystem,
+		RoleOthers:
 	}
 
-	r.filters = make([]Filter, len(v2filters))
-	for i := range v2filters {
-		r.filters[i] = *NewFilterFromV2(&v2filters[i])
+	for i := range r.target.roles {
+		if r.target.roles[i] == role {
+			return true
+		}
 	}
 
-	return r
+	return false
 }
 
-// Marshal marshals Record into a protobuf binary form.
-func (r *Record) Marshal() ([]byte, error) {
-	return r.ToV2().StableMarshal(nil), nil
+// TargetBinaryKeys returns binary-encoded public keys of subjects to which this
+// access rule is matched.
+//
+// Return value MUST NOT be mutated, make a copy first if needed.
+//
+// See also [Record.IsForKey].
+func (r Record) TargetBinaryKeys() [][]byte {
```

Comment:
is it ok to have bytes here but `neofscrypto.PublicKey` in the func a few lines below?